README
IBM::LoadLeveler version 0.07
=============================

This module makes the IBM LoadLeveler Data Access, Query, Submit & Workload management APIs available to Perl.

LoadLeveler is a product of IBM.  For more information, please refer to http://www.ibm.com/

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install
   make html

DEPENDENCIES

This module requires IBM LoadLeveler 3.1.0 and AIX 5.1

  Copying

    Copyright (c) 2002 Mike Hawkins. All rights reserved. This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself. 


Changes
Revision history for Perl extension IBM::LoadLeveler.

---   Sat Apr 19 13:05 2008
        - Fixed a problem with LL_ClassPreemptClass call trying to free a NULL pointer 

---   Mon Nov 19 21:55 2007
        - Modified Makefile.PL to be more CPAN testers friendly
        - Added query.t test file

1.07  Fri Sep 21 13:30 2007
        - Added support for version 3.4.1.0 & 3.4.2.1
	- Added Configuration API
	- Added Workload API function ll_move_job & ll_move_spool
	- Removed status value return from ll_make_reservation

1.06  Tue Sep 18 17:05 2007
	- Added a few more llapi.h files so some definitions are now correctly
	  in older versions.
	- Initialised some variables to stop warnings on plinux.
	- Simple testing done on plinux
	- Added code for ll_start_job_ext
	- Added code for ll_preempt_jobs
	- Added support for version 3.3.1.0+
	- Major rewrite to alleviate some problems with version support, 
          defs.h.in now included 
        - Added ll_cluster, ll_cluster_auth
	- Added ll_fair_share
        - Added $IBM::LoadLeveler::errObj and stopped returning errObj
        - Modified ll_error

1.05  Fri Aug 26 12:54:00 2005
	- Added Linux Support
	- Added support for LoadLeveler 3.3.0.0+
	- Added llqres and reservation examples
	- Changed how the export list is built from enums
	- commented out a chunk of code that was created in the initial build
          since I don't know what it's for. I'll remove it in a later version
	  if nobody complains.

1.04  Tue Apr  5 13:00:00 2005
	- Fixed LL_AdapterWindowList & LL_MachinePoolList.

1.03  Web Oct 20 10:20:00 2004
	- Fixed problem with ll_control wrappers
	- Checked to see if any updates needed for version 3.2.0.11
	- Removed NULL string warning in XS_unpack_charPtrPtr

1.02  Mon Sep 20 12:31:00 2004
	- Analyzed Versions 3.1.0.24 & 3.2.0.0 - 3.2.0.9 for changes
	  updated LoadLeveler.pod and pm/xs files accordingly
	- exported symbols for ll_modify

1.01  Mon Aug 23 13:30:00 2004  
        - ll_reset_request not exported
        - bugfix from Bernd Kallies (kallies-AT-zib.de)
	- Stripped leading 0 from LLVER variable in Makefile
	  stops it looking like octal, which is bad for version 3.2.0.8
	- Fixed a few typos
	- Updated examples xmlrss-llq and rss-llq

1.00  Tue Jan 06 10:13:00 2004
	- Added support for LoadLeveler 3.2
		- LL_CLASS_* in Data Access API
		- Extra arguments to ll_modify
	- using lslpp in Makefile to work out a version number
	- More examples from Mark Roberts (mark.roberts-AT-awe.co.uk)

1.00  Fri Aug 29 15:34:00 2003
	- Added QUERY_PROCID parameter from APAR IY35212
		(Currently broken see APAR IY48329)
	- Added SYSTEM_PREEMPT_STEP from APAR IY37022
	- Fixed some problems with 64 bit compile

0.08  Mon Jun 09 16:18:00 2003
	- Split documentation into sections for easier maintenance
	- Removed some of the more innovative spellings
	- Added example of monitor program in Submit.pod

0.07  Tue Mar 18 09:44:00 2003
	- Changed module name from LoadLeveler to IBM::LoadLeveler for CPAN

0.06  Thu Mar 06 16:16:00 2003
	- Added Notes on Core dumps with large history files and solution 
	  contributed by Dr Peter Mayes (peter_mayes-AT-uk.ibm.com) 
	  and Mark Roberts (mark.roberts-AT-awe.co.uk)
	- Modified documentation

0.05  Tue Jan 28 16:15:00 2003
	- Added Workload Management API
	- Added Data Access API test
	- More Documentation

0.04  Tue Jan 21 15:20:00 2003
	- 64 bit data types are now truncated if they overflow the Perl type
	- Added Generation and installation of llapi.ph to Makefile.PL
	- Added html target to convert pod to html

0.03  Thu Jan 16 09:26:32 2003
	- Added submit API
	- Added ll_control function from Workload Management API
	- Added ll_error

0.02  Tue Jan 07 16:12:05 2003
	- Fixed int64_t ll_get_data return types
	- added STATE_ constants to export list
	- added machine_job_query & wlmstat examples
	- Changed package name to LoadLeveler
	- Improved 64 bit types

0.01  Thu Dec 12 11:49:27 2002
	- original version; created by h2xs 1.21 with options
		-x -n LoadLeveler llapi.h



Makefile.PL
use 5.006;
use ExtUtils::MakeMaker;

my $LLVER="UNKNOWN";

$CC=$ExtUtils::MakeMaker::Config{"cc"};

#
# I hope this bit works out the base LoadLeveler version.
# Can't use the variables in the llapi.h file because the version number is the
# same for 3.1 & 3.2!  I am sure there is a very good reason for this, I don't know
# what it is, but i'm sure it must be a very good reason.
#

if ( $^O eq "linux" )
{
    $LoadL="/opt/ibmll/LoadL/full";
    # Quit if no llapi.h file (for CPAN testers)
    if ( ! -f "$LoadL/include/llapi.h")
    {
	print STDERR "Unable to find LoadLeveler include file\n";
	exit 0;
    }
    $ExtraLibs="-lstdc++"; # stops load failure with error undefined symbol: __gxx_personality_v0

    my $rpm = "";
    $rpm = "/usr/bin/rpm" if -f "/usr/bin/rpm";
    $rpm = "/bin/rpm"     if -f "/bin/rpm";
    my @packages=`$rpm -qa`;
    foreach $package (@packages)
    {
	next unless $package=~/LoadL.*-([\d.]+)-/;
#	print "$package ----> $1\n";
	$LLVER=$1;
    }
    if ( ! $LLVER eq "UNKNOWN" )
    {
	print STDERR "Unable to extract LoadLeveler version\n";
	exit 0;
    }

    # 64 bit required on PowerPC
    if ( ! defined $ExtUtils::MakeMaker::Config{"use64bitall"} && $ExtUtils::MakeMaker::Config{"archname"} =~ /ppc-linux/)
    {
    	print "I'm not totally sure but I think you are trying to build with a 32 bit version of Perl, the llapi is 64 bit\n";	
	print "This is probably going to end very badly. Sorry\n\n";
    }
}
elsif  ( $^O eq "aix" )
{
    $LoadL="/usr/lpp/LoadL/full";
    # Quit if no llapi.h file (for CPAN testers)
    if ( ! -f "$LoadL/include/llapi.h")
    {
	print STDERR "Unable to find LoadLeveler include file\n";
	exit 0;
    }
    $ExtraLibs="";

    #
    # The code extracts the package version using lslpp.
    #
    open(LLVER,"/usr/bin/lslpp -Ou -qlc LoadL.full |");
    $_=<LLVER>;
    $_=~/^[^:]+:[^:]+:([^:]+):/;
    close(LLVER);
    $LLVER=$1;
}
else
{
    # Return a value that CPAN testers might like
    # http://cpantest.grango.org/cgi-bin/pages.cgi?act=wiki-page&pagename=CPANAuthorNotes
    print STDERR "OS unsupported\n";
    print STDERR "LoadLevler is only supported on Linux and AIX. Found $OS\n";
    exit 0;
}


#
# The code converts the package version to a number, eg
# String is 3.1.0.16
# 	doubles any single digits:      03.01.00.16
#       removes the dots:               03010016
#	removes the leading 0 (octal):  3010016
# The final number is used in #ifdef statements
#
# The 1 while ..... construct repeatedly runs the doubling search. This is needed
# because if you match one .0. then the start for the next search is the first digit
# in the next number. See Page 155 of Camel V3. Alternativly just trust me that it works :)
#

1 while $LLVER=~s/\.(\d)\./.0$1./g;
$LLVER=~s/\.(\d)$/.0$1/;
$LLVER=~s/\.//g;
$LLVER=~s/^0(\d+)/$1/;

print "Building for version $LLVER on $^O\n\n";

process_defs();

WriteMakefile(
    'NAME'		=> 'IBM::LoadLeveler',
    'VERSION_FROM'	=> 'LoadLeveler.pm',
    'PREREQ_PM'		=> {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?    ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM    => 'LoadLeveler.pod', # retrieve abstract from module
       AUTHOR           => 'Mike Hawkins <mike\@pink-pit.com>') : ()),
    'PM'                => { 'llapi.ph'        => '$(INST_LIBDIR)/llapi.ph',
			     'LoadLeveler.pm'  => '$(INST_LIBDIR)/LoadLeveler.pm'
			   },		   
    'MAN3PODS'		=> { 'LoadLeveler.pod' => '$(INST_MAN3DIR)/LoadLeveler.$(MAN3EXT)',
			     'DataAccess.pod'  => '$(INST_MAN3DIR)/LoadLeveler::DataAccess.$(MAN3EXT)',
			     'Submit.pod'      => '$(INST_MAN3DIR)/LoadLeveler::Submit.$(MAN3EXT)',
			     'Query.pod'       => '$(INST_MAN3DIR)/LoadLeveler::Query.$(MAN3EXT)',
			     'Workload.pod'    => '$(INST_MAN3DIR)/LoadLeveler::Workload.$(MAN3EXT)',
			     'Reservation.pod' => '$(INST_MAN3DIR)/LoadLeveler::Reservation.$(MAN3EXT)',
			     'Error.pod'       => '$(INST_MAN3DIR)/LoadLeveler::Error.$(MAN3EXT)',
			     'Fairshare.pod'   => '$(INST_MAN3DIR)/LoadLeveler::Fairshare.$(MAN3EXT)',
			   },
    'LIBS'		=> ["-L$LoadL/lib -lllapi $ExtraLibs"], # e.g., '-lm'
    'DEFINE'		=> "-DLLVER=$LLVER",
    'INC'		=> "-I$LoadL/include",
    'OBJECT'		=> '$(O_FILES) ',
    'MYEXTLIB'          => 'ct.a',  # Fake Library to con MakeMaker into building C tests
    'CC'                => $CC,
    'macro'		=> { LoadL => "$LoadL" },
    'clean'             => { FILES => "llapi.ph ct.a ct/version ct/data_access ct/int_types" },
);
sub MY::postamble {
    my $postamble = <<'END';
llapi.ph: $(LoadL)/include/llapi.h
	( cd $(LoadL)/include ; h2ph -d /tmp llapi.h )
	cat /tmp/llapi.ph | sed -e /require.*/d > $(INST_LIB)/../../llapi.ph
	rm /tmp/llapi.ph
html: LoadLeveler.pm DataAccess.pod Reservation.pod Submit.pod Query.pod Workload.pod Error.pod Fairshare.pod LoadLeveler.pod
	pod2html --flush --noindex --podpath=. --htmlroot=. --outfile=DataAccess.html DataAccess.pod
	pod2html --noindex --podpath=. --htmlroot=. --outfile=Reservation.html Reservation.pod
	pod2html --noindex --podpath=. --htmlroot=. --outfile=Submit.html Submit.pod
	pod2html --noindex --podpath=. --htmlroot=. --outfile=Query.html Query.pod
	pod2html --noindex --podpath=. --htmlroot=. --outfile=Workload.html Workload.pod
	pod2html --noindex --podpath=. --htmlroot=. --outfile=Error.html Error.pod
	pod2html --noindex --podpath=. --htmlroot=. --outfile=Fairshare.html Fairshare.pod
	pod2html --noindex --podpath=. --htmlroot=. --outfile=LoadLeveler.html LoadLeveler.pod
ct.a:	ct/version ct/data_access ct/int_types
	ar r ct.a	
ct/version: ct/version.c
	$(CC) -o $@ $(CCFLAGS) $(INC) $@.c $(EXTRALIBS) 	
ct/data_access: ct/data_access.c
	$(CC) -o $@ $(CCFLAGS) $(INC) $@.c $(EXTRALIBS) 	
ct/int_types: ct/int_types.c
	$(CC) -o $@ $(CCFLAGS) $(INC) $@.c $(EXTRALIBS) 	
END
    $postamble;
}

sub process_defs
{

    my %enums=();
    my @values=();

    print "Building list of current ll_get_data specifications\n";
    open(LLAPI,"$LoadL/include/llapi.h") || die "Can't open $LoadL/include/llapi.h $!";
    # Speed through the file till we find
    # the LLAPI_Specification enum
    while ( <LLAPI> )
    {
	last if /LLAPI_Specification/;
    }
    my $count=0;
    while ( <LLAPI> )
    {
	last if /}/;

	next if /Unused_LL/;
	next unless /LL/;
	chomp;
	# Don't bother with anything that doesn't have a comma in it.
	next unless /([^,]*),(.*)/;

	my $before=$1;
	my $after=$2;

	# If the version of C we are using does not support Designated Initializers
	# we need the enum id number. boo hiss btw
	$count = $1 if /=\s*(\d+),/;

	# Process the enum
	# remove leading whitespace
	$before=~s/^\s+//;
	# remove everything after a white space
	$before=~s/\s+.*//;
	# remove everything after and including an =
	$before=~s/=.*//;
	
	# Process the definition

	$after="NONE" if $after=~/^\s*$/;

	# remove anything up to the start of a type def
	$after=~s|^.*/\*\s*||;
	# remove everything after a :
	$after=~s|:.*||;
	# or a (
	$after=~s|\(.*||;
	# remove anything after a *
	$after=~s|\*\s+.*|\*|;
	# strip out C end comments
	$after=~s|\*/||;
	$after= uc($after);
	$after=~s/\s*$//g;
	$after=~s/\*/_STAR/g;
	$after=~s/ /_/g;
	$after=~s/_+/_/g;
	$after="LL_$after";
#	print "$count $before  -> $after  ($_)\n";

	$enums{$before}=$after;
	$values[$count++]=$before;
    }
    close(LLAPI);

    print "Generating new definitions array\n";
    open(DEFS,"defs.h.in") || die "Can't open defs.h.in $!";
    open(OUT,"> defs.h.new")|| die "Can't open defs.h.new $!";

    #
    # C99 introduced Designated Initalizers for arrays, meaning you
    # can do things like:
    # static int defs[]={
    #       [LL_StepCheckpointing] = LL_BOOLEAN_STAR,
    #       [LL_MachineName] = LL_CHAR_STAR,
    #                   };
    # Which is very nice since the alternative is to build a very nasty array.
    #
    # For now I don't know how to work out if C supports this feature, I do know it
    # doesn't work on my version of AIX though...
    #
    my $UseDesignatedInitializers = 1;

    $UseDesignatedInitializers = 0 if $OS =~ /AIX/;

    # Copy everything out until we get to the array definition
    while( <DEFS> )
    {
	print OUT $_;
	last if ( $_ =~ /\[\]/ );
    }

    #
    # We are now pointing at the definitions in the template array
    # What I want to do at this point is to read the defs.h.in file
    # and if I find a line like
    #       [LL_StepCheckpointing] = LL_BOOLEAN_STAR,
    # then see if the LL_ thingie was defined in the
    # include file, if it is write it out, if not don't.
    #
    # However this will only work if the Designated Initializers
    # are supported, otherwise a fully defined array needs to be built
    #
    # So, suck my handcranked defs in from the defs.h.in file
    # and then loop through all enums as slurped out of the
    # system llapi.h file, @values contains a number to LL_ thingie
    # which can be used to see if there is anything to print.
    #

    my %defs=();
    while ( <DEFS> )
    {
	last if ( $_ !~ /\[.*\]/ );
	# slurp the definition from the file into a hash
	next unless $_=~/\[(.*)\]\s*=\s*(\S+)\s*\,/;
	$defs{$1}=$2;
    }

    for(my $i=0;$i<$count;$i++)
    {
	if ( exists $defs{$values[$i]})
	{
	    if ( $UseDesignatedInitializers)
	    {
		print OUT "\t\[$values[$i]\] = $defs{$values[$i]},\n";
	#	print "\t\[$values[$i]\] = $defs{$values[$i]},\n";
	    }
	    else
	    {
		print OUT "$defs{$values[$i]},\n";
	    }
	    delete $enums{$values[$i]};
	    delete $defs{$values[$i]};
	}
	else
	{
	    print OUT "LL_NONE,\n" if ! $UseDesignatedInitializers;
	}
    }

    # Close the array
    print OUT "};\n";
    print OUT "\n";
    close(OUT);
    close(DEFS);


    if  ( scalar keys %enums > 0 )
    {
	print "Current llapi.h has the following definitions, add them to defs.h at your own risk:\n\n";
	foreach my $key ( keys %enums)
	{
	    print "[$key] = $enums{$key},\n";
	}
	print "\n";
    }
    if  ( scalar keys %defs > 0 )
    {
	print "defs.h.in has the following definitions not seen in your system llapi.h file.  They have been removed from defs.h:\n\n";
	foreach my $key ( keys %defs)
	{
	    print "[$key] = $defs{$key},\n";
	}
	print "\n";
    }

    if ( -s "defs.h.new" )
    {
	unlink "defs.h.old" if ( -f "defs.h.old");
	rename "defs.h","defs.h.old" if ( -f "defs.h" );
	rename "defs.h.new","defs.h";
    }
}
